src/cluster_migrate.c

```diff
@@ -0,0 +1,1953 @@
+        return;
+    }
+    updateSlotMigrationLinkState(link, SLOT_EXPORT_SNAPSHOTTING);
+    if (server.debug_pause_after_fork) debugPauseProcess();
+}
+
+/* Callback triggered after snapshot is finished. We either begin sending the incremental contents
+ * or fail the associated migration. */
+void clusterHandleSlotExportBackgroundSaveDone(int bgsaveerr) {
+    listIter li;
+    listNode *ln;
+    listRewind(server.cluster->slot_migration_links, &li);
+    while ((ln = listNext(&li))) {
+        slotMigrationLink *link = (slotMigrationLink *)ln->value;
+        if (link->type != SLOT_MIGRATION_EXPORT) continue;
+        if (link->state != SLOT_EXPORT_SNAPSHOTTING) {
+            continue;
+        }
+        if (bgsaveerr == C_OK) {
+            serverLog(LL_NOTICE,
+                      "Finished snapshotting slot migration %s, beginning incremental stream...",
+                      link->description);
+            slotExportBeginStreaming(link);
+        } else {
+            serverLog(LL_WARNING,
+                      "Child process failed to snapshot slot migration %s",
+                      link->description);
+            finishSlotMigrationLink(link, SLOT_MIGRATION_LINK_FAILED, "Failed to perform snapshot");
+        }
+    }
+}
+
+/* Returns 1 if the attached slot migration link for the client is ready to flush the data from the
+ * output buffer. Returns 0 otherwise. While this returns 0, replicated data for this slot migration
+ * will accumulate in the client output buffer. */
+int clusterSlotMigrationShouldInstallWriteHandler(client *c) {
+    slotMigrationLink *link = (slotMigrationLink *)c->slot_migration_link;
+    if (!link || link->type != SLOT_MIGRATION_EXPORT) {
+        return 1;
+    }
+    return link->state != SLOT_EXPORT_SNAPSHOTTING;
+}
+
+/* Feed the slot export links with the given argv and argc from the replication
+ * log. Slot is pre-populated using the original command that was run. */
+void clusterFeedSlotExportLinks(int dbid, robj **argv, int argc, int slot) {
+    UNUSED(dbid);
+
+    if (slot == -1) {
+        /* We can safely ignore any commands with no keys. This includes
+         * MULTI/EXEC. This isn't a problem since the entire slot migration is
+         * atomically visible and therefore transactions are redundant. */
+        return;
+    }
+
+    listIter li;
+    listNode *ln;
+    listRewind(server.cluster->slot_migration_links, &li);
+
+    /* Select cmd will be generated only if at least one client needs it. */
+    robj *selectcmd = NULL;
+    while ((ln = listNext(&li))) {
+        slotMigrationLink *link = (slotMigrationLink *)ln->value;
+        if (link->type != SLOT_MIGRATION_EXPORT) continue;
+        if (!link->client) continue;
+        if (!isSlotMigrationLinkInProgress(link) || link->state < SLOT_EXPORT_SNAPSHOTTING) continue;
+        if (!isSlotInSlotRanges(slot, link->slot_ranges)) continue;
```

Comment:
`CLUSTER FLUSHSLOT` is odd:
- On the exporting node, it is not propagated, as it has no slot.
- On the importing node, a client is able to flush a slot that is importing

Both lead to inconsistent data after migration.